cvat-ui/src/components/common/preview.tsx

```diff
@@ -45,55 +45,47 @@ export default function Preview(props: Props): JSX.Element {
         previewClassName,
     } = props;
 
+    const [hasFetched, setHasFetched] = React.useState(false);
+    const { ref, inView } = useInView({ triggerOnce: true });
+
     const preview = useSelector((state: CombinedState) => {
-        if (job !== undefined) {
-            return state.jobs.previews[job.id];
-        } if (project !== undefined) {
-            return state.projects.previews[project.id];
-        } if (task !== undefined) {
-            return state.tasks.previews[task.id];
-        } if (cloudStorage !== undefined) {
-            return state.cloudStorages.previews[cloudStorage.id];
-        } if (model !== undefined) {
-            return state.models.previews[model.id];
-        }
+        if (job !== undefined) return state.jobs.previews[job.id];
+        if (project !== undefined) return state.projects.previews[project.id];
+        if (task !== undefined) return state.tasks.previews[task.id];
+        if (cloudStorage !== undefined) return state.cloudStorages.previews[cloudStorage.id];
+        if (model !== undefined) return state.models.previews[model.id];
         return '';
     });
 
     useEffect(() => {
-        if (preview === undefined) {
-            if (job !== undefined) {
-                dispatch(getJobPreviewAsync(job));
-            } else if (project !== undefined) {
-                dispatch(getProjectsPreviewAsync(project));
-            } else if (task !== undefined) {
-                dispatch(getTaskPreviewAsync(task));
-            } else if (cloudStorage !== undefined) {
-                dispatch(getCloudStoragePreviewAsync(cloudStorage));
-            } else if (model !== undefined) {
-                dispatch(getModelPreviewAsync(model));
-            }
+        if (inView && !hasFetched && preview === undefined) {
+            setHasFetched(true);
+            if (job !== undefined) dispatch(getJobPreviewAsync(job));
+            else if (project !== undefined) dispatch(getProjectsPreviewAsync(project));
+            else if (task !== undefined) dispatch(getTaskPreviewAsync(task));
+            else if (cloudStorage !== undefined) dispatch(getCloudStoragePreviewAsync(cloudStorage));
+            else if (model !== undefined) dispatch(getModelPreviewAsync(model));
         }
-    }, [preview]);
+    }, [inView, hasFetched, preview]);
 
     if (!preview || (preview && preview.fetching)) {
         return (
-            <div className={loadingClassName || ''} aria-hidden>
+            <div ref={ref} className={loadingClassName || ''} aria-hidden>
                 <Spin size='default' />
             </div>
         );
     }
 
     if (preview.initialized && !preview.preview) {
         return (
-            <div className={emptyPreviewClassName || ''} onClick={onClick} aria-hidden>
+            <div ref={ref} className={emptyPreviewClassName || ''} onClick={onClick} aria-hidden>
```

Comment:
Is there a reason to put this ref here and below? (preview already downloaded)

Reply:
Removed